### PR TITLE
Oneandzero24/settings

### PR DESCRIPTION
**Pull request by OneAndZero24**
Implemented new settings:
- Domain Incremental
- Task Incremental

### CODE REVIEW

**Review by pkrukowski1 — CHANGES_REQUESTED** [review]
Very good job! The Domain-Incremental scenario is well implemented. However, there are issues with Task-Incremental Learning because:
- In EWC, MAS, and SI, we should store one set of importance parameters per task and then regularize weight updates using the correct set of importance parameters.
- In LwF, where we operate on multiple heads, we should calculate the softmax separately for each head rather than for the entire neural network’s output.

**Comment by OneAndZero24** [comment]
> Very good job! The Domain-Incremental scenario is well implemented. However, there are issues with Task-Incremental Learning because:
> 
> * In EWC, MAS, and SI, we should store one set of importance parameters per task and then regularize weight updates using the correct set of importance parameters.
> * In LwF, where we operate on multiple heads, we should calculate the softmax separately for each head rather than for the entire neural network’s output.

Please note how TI is implemented in composer especially `composer.py:97`. Heads are stored in regular list not `nn.Modulelist` what's more in `composer.py:147-153` we can see we are actually replacing head in `module.classifier`. During training/inference on a single task other heads/parameters from them are invisible to pytorch/our methods which are interfacing with `method.model`. I've checked it with listing `.children()` and `.parameters` inside MAS, EWC, SI, LwF. It should mean none of the problems mentioned above occur as each plugin stores head parameters and calculates importance in setup task where head is already replaced. LwF wont see any other heads than current task one.
If it answears your doubts please mark as resolved.

**Comment by pkrukowski1** [comment]
> > Very good job! The Domain-Incremental scenario is well implemented. However, there are issues with Task-Incremental Learning because:
> > 
> > * In EWC, MAS, and SI, we should store one set of importance parameters per task and then regularize weight updates using the correct set of importance parameters.
> > * In LwF, where we operate on multiple heads, we should calculate the softmax separately for each head rather than for the entire neural network’s output.
> 
> Please note how TI is implemented in composer especially `composer.py:97`. Heads are stored in regular list not `nn.Modulelist` what's more in `composer.py:147-153` we can see we are actually replacing head in `module.classifier`. During training/inference on a single task other heads/parameters from them are invisible to pytorch/our methods which are interfacing with `method.model`. I've checked it with listing `.children()` and `.parameters` inside MAS, EWC, SI, LwF. It should mean none of the problems mentioned above occur as each plugin stores head parameters and calculates importance in setup task where head is already replaced. LwF wont see any other heads than current task one. If it answears your doubts please mark as resolved.

OK, I have no more doubts. Once again - very good job!

**Review by pkrukowski1 — APPROVED** [review]
Great job!
